### PR TITLE
Static Components + Primitives 

### PR DESCRIPTION
**Pull request by calebmkim**
Adds Static Components + Primitives + Static Invokes to the IR, by adding a `latency` field to components and `CellType::Primitive`. 

Few notes: 

1) In order to infer the latency of `static invokes` (i.e., when we see `static invoke static_comp()()` we infer the latency rather than forcing the Calyx user to write `static<latency> ...`) we have to do build components in `from_ast` in post order (bc we want to be able to infer the static latency of components too). This is similar to the `PostOrder` thing for `Visitor`. For this reason, the ordering the components is printed is changed and I had to overwrite a few of the tests. 

2) I'm not quite sure what the requirements should be for when we can use static components/primitives. Currently, here is what we enforce: 

- For non-static groups (i.e, comb groups and dynamic groups), you are not allowed to write to static primitives or static components. I honestly am leaning towards changing this (i.e., allow dynamic, non-comb groups to write to static components). My reasoning is that when we do `infer-static-timing` and upgrade a component from dynamic->static, we would have to rewrite every dynamic group that writes to that component to be a static group, which could be challenging (maybe even impossible). 

(We also enforce these other ones) 
- Static invokes must be of static components/primitives (clearly we must enforce this) 
- Non static invokes must be of non-static components/primitives (not 100% sure if we should enforce this... is there a case where we want to dynamically invoke primitives?)

### CODE REVIEW

File: calyx-frontend/src/ast.rs [hunk 1]
```diff
@@ -63,10 +63,22 @@ pub struct ComponentDef {
     pub attributes: Attributes,
     /// True iff this is a combinational component
     pub is_comb: bool,
+    /// True iff this is a static component
+    pub is_static: bool,
+    /// (Optional) latency of component, if it is static
+    /// NOTE: latency can still be None even if is_static is true. That
+    /// just means the compiler will infer the latency
+    pub latency: Option<NonZeroU64>,
```
Comment (rachitnigam) [inline calyx-frontend/src/ast.rs:71]: Hm, this is an interesting design choice. I was thinking that we only have the `latency` field and nothing else. I'm not sure what having `is_static` in addition to that accomplishes. Is the idea that there are three kinds of components:
1. Static components with specified latency
2. Static components without specified latency
3. Dynamic components

If that is the case, then (2) & (3) should probably be same for the following reason: Using the `static` keyword on a component is declaring a calling convention–it says that it is not safe to call this component using a dynamic invoke. In contrast, a component that merely *happens* to use only `static` groups and control is not declaring its interface as `static` and the compiler is free to choose whether it is promoted to a static component or compiled into a dynamic one.

File: calyx-frontend/src/parser.rs [hunk 2]
```diff
@@ -1029,71 +1040,111 @@ impl CalyxParser {
...
+                    continuous_assignments,
+                    control,
+                    attributes: attributes.add_span(span),
+                    is_comb: false,
+                    is_static: false,
+                    latency: None,
+                })
+            },
+            [
+                static_optional_latency(latency),
```
Comment (rachitnigam) [inline calyx-frontend/src/parser.rs:1113]: What happens if someone specifies both `comb` and `static`? We should probably generate a reasonable here in the parser using `input.error` method and say that comb and static at the same time are not allowed. Also, worth moving the logic of parsing the keywords in front of a `component` into a separate method to deduplicate some of this stuff.

File: calyx-frontend/src/syntax.pest [hunk 3]
```diff
@@ -103,15 +107,15 @@ sig_with_params = {
   params? ~ signature
 }
 primitive = {
-     comb? ~ "primitive" ~ name_with_attribute ~ sig_with_params ~ ";"
+     (comb | static_annotation)? ~ "primitive" ~ name_with_attribute ~ sig_with_params ~ ";"
```
Comment (rachitnigam) [inline calyx-frontend/src/syntax.pest:110]: Moving the `(comb | static_annotation)` stuff into a new rule would help with the thing I mentioned in `parser.rs`

File: calyx-ir/src/component.rs [hunk 4]
```diff
@@ -41,6 +41,8 @@ pub struct Component {
     pub attributes: Attributes,
     /// True iff component is combinational
     pub is_comb: bool,
+    /// (Optional) latency of component, if it is static
+    pub latency: Option<u64>,
```
Comment (rachitnigam) [inline calyx-ir/src/component.rs:45]: Why not use a `NonZeroU64` here as well?

File: calyx-opt/src/passes/well_formed.rs [hunk 5]
```diff
@@ -505,6 +563,64 @@ impl Visitor for WellFormed {
...
+        let ports: HashSet<_> =
+            cell.ports.iter().map(|p| p.borrow().name).collect();
+
+        s.inputs
+            .iter()
+            .chain(s.outputs.iter())
+            .try_for_each(|(port, _)| {
+                if !ports.contains(port) {
+                    Err(Error::malformed_structure(format!(
+                        "`{}` does not have port named `{}`",
```
Comment (rachitnigam) [inline calyx-opt/src/passes/well_formed.rs:584]: Can this error ever happen? I thought `from_ast` would take care of ensuring that all ports are defined named on a particular invoke.

File: primitives/pipelined.futil [hunk 6]
```diff
@@ -1,6 +1,6 @@
 extern "pipelined.sv" {
     // A latency-sensitive multiplier that takes 4 cycles to compute its result.
-    primitive pipelined_mult<"latency"=4>(
+    static<4> primitive pipelined_mult (
```
Comment (rachitnigam) [inline primitives/pipelined.futil:3]: Awesome! CC @andrewb1999

**Comment by rachitnigam** [comment]
Thanks for getting us started! The big design decision here that is subject to discussion is whether we should allow `static` components without a specified latency. I think the answer is maybe, especially once we start thinking about `virtual` operators. However, I'm not sure if having `from_ast` infer the latencies of the component and propagate them to `static invoke` is the right thing. Can you talk a bit more about this decision and what the trade-offs were?

**Comment by calebmkim** [comment]
The reason for wanting to infer the latencies of `static_component` is the same reason we infer latencies of `static seq`: it is just cleaner and more convenient to not have to change the latency annotations every time you make a small control change. However, if a frontend is generating this code (which is most of the time), it doesn't seem like this is a very strong reason. 

Also, just to be clear, by the time we get `ir::Component`'s, all static components will have a latency (i.e., there is no `is_static` field on an `ir::Component`. We just look at the  latency` field to check if it's static or not. It's just in the `ast::ComponentDef`s that have the weird fields), so in the IR, it's _as if_  we specified a latency on each static component. Also, it's entirely possible to have a dynamic component that only uses static control: the person just has to not include the `static` keyword on the component declaration. 

The alternative would  making all static components to specify a latency (we need some way of knowing what the latency is of static components so we can assign them a latency when they're invoked). That would actually simplify the `from_ast` and the `parser` code dramatically. It would also fit nicely with the comparison that groups:enables as components:invokes. We would require explicit latency annotations on (static) groups and components, but not on (static) enables and invokes. 

In other words, there aren't necessarily strong reasons to stick with what I have, and changing it would actually simplify the code.

**Comment by rachitnigam** [comment]
Right, that's what I thought too! I also think the whole philosophical argument for `static` being a calling convention is a good reason to keep things explicit now. We will have to revisit this decision when we implement virtual operators since it could be the case that a component is `static` but it needs its virtual operators resolved before it can be fully assigned a latency

File: calyx-opt/src/passes/well_formed.rs (hunk 5 above)
Comment (calebmkim) [reply calyx-opt/src/passes/well_formed.rs:577]: I don't think it does, although we could change it. To do it would require passing the `SigCtx` for the program as an argument when building the control of the component.

Comment (calebmkim) [reply calyx-opt/src/passes/well_formed.rs]: ok, changed it.

**Comment by rachitnigam** [comment]
Also, I'll be asleep soon so feel free to merge if it unblocks progress on the rest of the static push.

**Comment by calebmkim** [comment]
ok merging. @paili0628 hopefully this can unblock you a bit.

**Comment by sampsyo** [comment]
> For non-static groups (i.e, comb groups and dynamic groups), you are not allowed to write to static primitives or static components. I honestly am leaning towards changing this (i.e., allow dynamic, non-comb groups to write to static components). My reasoning is that when we do `infer-static-timing` and upgrade a component from dynamic->static, we would have to rewrite every dynamic group that writes to that component to be a static group, which could be challenging (maybe even impossible).

This reasoning makes sense to me. It is like what I tried to outline in https://github.com/cucapra/calyx/issues/1417#issuecomment-1538922405. To summarize, I don't see any actual problem with dynamic groups using static components. Dynamic groups can mostly do anything anyway!

> Non static invokes must be of non-static components/primitives (not 100% sure if we should enforce this... is there a case where we want to dynamically invoke primitives?)

Also seems right to me. It's nice to keep these clearly distinct, to avoid invisibly implying extra "wrapper" hardware, unless we really need this kind of mixing for some reason.

**Comment by sampsyo** [comment]
Interesting discussion about whether we should have static stuff (in general!) without a latency. To try it on for size, an alternative policy would be: wherever you would want to do this, just use the dynamic thing. That is, if you have a component that you _want_ to be static but don't know its latency, too bad—you must make it a dynamic component. It is then the compiler's job to "upgrade" it to static once it determines the latency.

In other words, this policy would equate "unknown latency" with "dynamic" and "known latency" with "static." There would be no such thing as wishfully stating that you would _like_ the latency to be known but don't; you have to honestly fess up and say you don't know the latency for now and therefore the thing has to be dynamic for now.

This would fit nicely with our story about the compiler's main job in Calyx 2.0 stuff being to transparently and opportunistically "upgrade" stuff from dynamic to static while preserving its semantics.

Does this make any sense?

**Comment by calebmkim** [comment]
> In other words, this policy would equate "unknown latency" with "dynamic" and "known latency" with "static." There would be no such thing as wishfully stating that you would like the latency to be known but don't; you have to honestly fess up and say you don't know the latency for now and therefore the thing has to be dynamic for now.
This would fit nicely with our story about the compiler's main job in Calyx 2.0 stuff being to transparently and opportunistically "upgrade" stuff from dynamic to static while preserving its semantics.
Does this make any sense?

Yep it makes sense, and that's actually what this PR ended up implementing. It requires you to explicitly annotate the latency  if you want to declare a component static .

> This reasoning makes sense to me. It is like what I tried to outline in https://github.com/cucapra/calyx/issues/1417#issuecomment-1538922405. To summarize, I don't see any actual problem with dynamic groups using static components. Dynamic groups can mostly do anything anyway!

Yeah, I also changed this to allow static components anywhere. _However_ the PR does check that static primitives cannot be used in non-static groups. However, it would be very easy to change this if we want to, and after thinking about it more think I may be leaning to changing this too.

**Comment by sampsyo** [comment]
Awesome!! Yeah, in general, it seems nice for there to be no distinction between primitives and Calyx-implemented components on the "client" side, I suppose...

**Comment by rachitnigam** [comment]
Yup! I think the policy is quite sane and we can treat the process of removing `virtual` operators as a pre-step to the control and component promotion logic; once all the `virtual` ops have been removed, it a component can be promoted to a `static` component then we'll do so. Otherwise, it'll stay a dynamic component!
